man/chpobox.1
.TH CHPOBOX 1 "5 Feb 1988" "Project Athena"
\" RCSID: $HeadURL: svn+ssh://svn.mit.edu/moira/trunk/moira/man/chpobox.1 $ $Id: chpobox.1 3956 2010-01-05 20:56:56Z zacheiss $"
.SH NAME
chpobox \- change address at which user receives mail
.SH SYNOPSIS
.TP 8
chpobox [ username ]
.TP 8
chpobox [ \-s address | \-S address | \-p ] [ \-u username ]
.SH DESCRIPTION
Your post office box is what determines where you will receive
mail.  It can be of several types: 
.I POP, IMAP, SMTP, SPLIT
or
.IR NONE .

In the first use of this command, the post office box of the
user whose name corresponds to the single argument
.I username
(or your own name if 
.I username
is not specified) is displayed.

In the second use, this command sets the 
actual address that will receive mail sent to 
.I username@MIT.EDU
(where 
.I username
is your own name if you have not specified it on the command line).
It should be used if you want to change where you will 
receive mail.  

The 
.I chpobox 
command is used instead of the file
.I .forward,
which would be used with standard UNIX mail. 

In order to change the address that receives mail sent to 
.I username@MIT.EDU,
use the 
.B -s
option.  Specifically, to reroute your mail to
the address 
.I name@host,
you would issue the command

.nf
.nj
chpobox -s name@host
.fi

To split mail between your local post office box and a remote address,
use the
.B -S
option.

To restore your post office box to its original Athena default,
use the 
.B -p
flag.  This will set your mailbox to the last POP or IMAP server
it was using.
.PP

Only one of 
.B -p
or
.B -s
or
.B -S
may be used at a time.

Note that the actual change may take up to a day to 
propogate to the mail hub.

.SH NOTE

When a mailbox is stored in the central database, any host mentioned
will have its name canonicalized to a standard form.  If for some
reason you don't want to have the hostname canonicalized, put
.B just the hostname
in double-quotes.  For example, the command
.nf
.nj
	chpobox -s jrhacker@\\"cd.mit.edu\\"
.fi
will not be canonicalized to jrhacker@ls.mit.edu, as it would
otherwise.  Also note that the double-quotes were preceded by
backslashes to get them past the shell.

.SH SEE ALSO
moira(1),

Moira Section of the Project Athena Technical Plan.


debian/changelog
debathena-moira (4.0.0+svn20100225-0debathena2) UNRELEASED; urgency=low

  * There is no "lib" section - it's "libs".

 -- Evan Broder <broder@mit.edu>  Fri, 26 Feb 2010 12:50:19 -0500

debathena-moira (4.0.0+svn20100225-0debathena1) unstable; urgency=low

  * New SVN snapshot.

 -- Evan Broder <broder@mit.edu>  Thu, 25 Feb 2010 15:54:26 -0500

debathena-moira (4.0.0+svn20100223-0debathena1) unstable; urgency=low

  * New SVN snapshot.
  * Drop patch install-headers, which was incorporated upstream.

 -- Evan Broder <broder@mit.edu>  Thu, 25 Feb 2010 02:13:44 -0500

debathena-moira (4.0.0+svn20100104-0debathena1) unstable; urgency=low

  * New SVN snapshot.
  * Build and install libmrclient as a shared library.
  * Deal with rename of /etc{/athena => }/moira.conf
  * Patch manpages to deal with our installation paths.

 -- Evan Broder <broder@mit.edu>  Mon, 04 Jan 2010 22:05:10 -0500

debathena-moira (4.0.0+cvs20091116-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Drop patch fix-update-ticket, which was incorporated upstream.
  * Add a missing definition of ${prefix} in include/Makefile.in so that
    the moira.h header actually gets installed in
    /usr/include/moira. (Trac: #445)

 -- Evan Broder <broder@mit.edu>  Thu, 19 Nov 2009 10:53:40 -0500

debathena-moira (4.0.0+cvs20090922-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Add patch to fix builds without krb4.

 -- Evan Broder <broder@mit.edu>  Tue, 22 Sep 2009 00:35:12 -0400

debathena-moira (4.0.0+cvs20090728-0debathena3) unstable; urgency=low

  * Add a missing conflicts to make upgrades run more smoothly.

 -- Evan Broder <broder@mit.edu>  Tue, 28 Jul 2009 18:28:33 -0700

debathena-moira (4.0.0+cvs20090728-0debathena2) unstable; urgency=low

  * Include the headers in debathena-libmoira-dev.
  * Install update_test in moira-clients instead of moira-update-server.

 -- Evan Broder <broder@mit.edu>  Tue, 28 Jul 2009 17:54:34 -0700

debathena-moira (4.0.0+cvs20090728-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Fix zephyr dependencies to actually pull in Kerberized libraries; drop
    build-dep on Kerberized library.

 -- Evan Broder <broder@mit.edu>  Tue, 28 Jul 2009 13:43:28 -0400

debathena-moira (4.0.0+cvs20090621-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Drop krb4less-fix; it was incorporated upstream.
  * Universally stop building Moira with krb4 support.

 -- Evan Broder <broder@mit.edu>  Fri, 26 Jun 2009 00:51:28 -0400

debathena-moira (4.0.0+cvs20090618-0debathena1) unstable; urgency=low

  * New upstream release
  * Build and install libmoira as a shared library. (Trac: #70)
  * Drop most of the krb4 patch - it's been incorporated upstream.
  * Install the Moira development headers by patching the relevant
    Makefiles, instead of in the debian/rules file.

 -- Evan Broder <broder@mit.edu>  Sat, 20 Jun 2009 14:45:32 -0400

debathena-moira (4.0.0+cvs20090409-0debathena1) unstable; urgency=low

  * New CVS snapshot (Trac: #195)
  * Drop patches that have been incorporated upstream.
  * Add patch to allow building on systems without krb4. (Trac: #198)

 -- Evan Broder <broder@mit.edu>  Sun, 12 Apr 2009 14:03:22 -0400

debathena-moira (4.0.0+cvs20080830-0debathena2) unstable; urgency=low

  * Remove DEB_AUTO_UPDATE_DEBIAN_CONTROL.
  * Change "Debian-Athena Project" to "Debathena Project".
  * Remove debian/control from version control.

 -- Tim Abbott <tabbott@mit.edu>  Sat, 24 Jan 2009 15:32:35 -0500

debathena-moira (4.0.0+cvs20080830-0debathena1) unstable; urgency=low

  * New CVS snapshot
  * Add a debathena-moira-update-server package that installs
    update_server

 -- Evan Broder <broder@mit.edu>  Fri, 01 Aug 2008 10:54:26 -0400

debathena-moira (4.0.0+cvs20080701-0debathena1) unstable; urgency=low

  * New CVS snapshot
  * Move Moira header files from /usr/share/include to /usr/include/moira

 -- Evan Broder <broder@mit.edu>  Tue, 01 Jul 2008 20:38:13 -0400

debathena-moira (4.0.0+cvs20080304-0debathena1) unstable; urgency=low

  * New CVS snapshot to fix chsh.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 04 Mar 2008 15:50:31 -0500

debathena-moira (4.0.0+cvs20080301-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Remove mrtest segfault patch from last release (fixed in upstream)

 -- Tim Abbott <tabbott@mit.edu>  Sat,  1 Mar 2008 18:51:40 -0500

debathena-moira (4.0.0+cvs20070817-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Patch an mrtest segfault with too many command line arguments.

 -- Anders Kaseorg <andersk@mit.edu>  Fri, 24 Aug 2007 19:53:49 -0400

debathena-moira (4.0.0+cvs20070129-0debathena3) unstable; urgency=low

  * Move chsh.debathena to chsh.moira and chfn.debathena to chfn.moira.

 -- Anders Kaseorg <andersk@mit.edu>  Mon, 06 Aug 2007 03:17:10 -0400

debathena-moira (4.0.0+cvs20070129-0debathena2) unstable; urgency=low

  * Add missing build dependencies to libmoira-dev.
  * Section: debathena/net.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 17 Jul 2007 06:34:07 -0400

debathena-moira (4.0.0+cvs20070129-0debathena1) unstable; urgency=low

  * Initial release.

 -- Anders Kaseorg <andersk@mit.edu>  Wed, 14 Feb 2007 00:53:24 -0500


clients/chpobox/chpobox.c
/* $Id: chpobox.c 3956 2010-01-05 20:56:56Z zacheiss $
 *
 * Talk to the Moira database to change a person's home mail machine. This may
 * be an Athena machine, or a completely arbitrary address.
 *
 * chpobox with no modifiers reports the current mailbox.
 *
 * chpobox -s address means set the mailbox to this address.
 *
 * chpobox -p restores the pobox to a previous POP/IMAP box, if there was one.
 *
 * chpobox -S address means split mail between POP/IMAP and SMTP
 *
 * chpobox -u [user] is needed if you are logged in as one user, but
 * are trying to change the email address of another.  You must have
 * Kerberos tickets as the person whose address you are trying to
 * change, or the attempt will fail.
 *
 * Copyright (C) 1987-1998 by the Massachusetts Institute of Technology
 * For copying and distribution information, please see the file
 * <mit-copyright.h>.
 */

#include <mit-copyright.h>
#include <moira.h>
#include <moira_site.h>
#include <mrclient.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#define argis(a, b) (!strcmp(*arg + 1, a) || !strcmp(*arg + 1, b))

RCSID("$HeadURL: svn+ssh://svn.mit.edu/moira/trunk/moira/clients/chpobox/chpobox.c $ $Id: chpobox.c 3956 2010-01-05 20:56:56Z zacheiss $");

int get_pobox(int argc, char **argv, void *callarg);
void usage(void);

char *whoami;

static int match;

int main(int argc, char *argv[])
{
  char *mrarg[3];
  char *address, *uname;
  char **arg = argv;
  char *server = NULL;
  int c, setflag, splitflag, prevflag, status;

  setflag = splitflag = prevflag = 0;
  address = uname = NULL;

  if ((whoami = strrchr(argv[0], '/')) == NULL)
    whoami = argv[0];
  else
    whoami++;

  if (argc > 5)
    usage();

  while (++arg - argv < argc)
    {
      if (**arg == '-')
	{
	  if (argis("s", "set")) {
	    if (arg - argv < argc - 1) {
	      arg++;
	      setflag++;
	      address = *arg;
	    } else
	      usage();
	  }
	  else if (argis("S", "split")) {
	    if (arg - argv < argc - 1) {
	      arg++;
	      splitflag++;
	      address = *arg;
	    } else
	      usage();
	  }
	  else if (argis("p", "previous"))
	    prevflag++;
	  else if (argis("u", "username")) {
	    if (arg - argv < argc - 1) {
	      arg++;
	      uname = *arg;
	    } else
	      usage();
	  }
	  else if (argis("db", "database")) {
	    if (arg - argv < argc - 1) {
	      arg++;
	      server = *arg;
	    } else
	      usage();
	  }
	}
      else if (uname == NULL)
	uname = *arg;
      else
	usage();
    }

  if (prevflag + splitflag + setflag > 1)
    usage();

  if (!uname)
    {
      uname = mrcl_krb_user();
      if (!uname)
	exit(1);
    }
  mrarg[0] = uname;

  if (mrcl_connect(server, "chpobox", 2, 1) != MRCL_SUCCESS)
    exit(1);

  if (setflag || splitflag)
    {
      char *addr;
      status = mrcl_validate_pobox_smtp(uname, address, &addr);
      if (mrcl_get_message())
	mrcl_com_err(whoami);
      if (status != MRCL_SUCCESS)
	{
	  printf("\n");
	  goto show;
	}
      mrarg[1] = setflag ? "SMTP" : "SPLIT";
      mrarg[2] = addr;
      status = mr_query("set_pobox", 3, mrarg, NULL, NULL);
      free(addr);
      if (status)
	{
	  com_err(whoami, status,
		  "while setting pobox for %s to type %s, box %s",
		  mrarg[0], mrarg[1], mrarg[2]);
	}
    }
  else if (prevflag)
    {
      status = mr_query("set_pobox_pop", 1, mrarg, NULL, NULL);
      if (status == MR_MACHINE)
	{
	  fprintf(stderr,
		  "Moira has no record of a previous POP box for %s\n", uname);
	}
      else if (status != 0)
	com_err(whoami, status, "while setting pobox");
    }

  /*
   * get current box
   */
show:
  status = mr_query("get_pobox", 1, mrarg, get_pobox, NULL);
  if (status == MR_NO_MATCH)
    printf("User %s has no pobox.\n", uname);
  else if (status != 0)
    mrcl_com_err(whoami);
  mr_disconnect();
  exit(0);
}


/*
 * get_pobox gets all your poboxes and displays them.
 */

int get_pobox(int argc, char **argv, void *callarg)
{
  if (!strcmp(argv[1], "SMTP"))
    {
      printf("User %s, Type %s, Box: %s\n",
	     argv[0], argv[1], argv[2]);
    }
  else if (argc == 7)
    {
      printf("User %s, Type %s, Box: %s (%s)\n",
	   argv[0], argv[1], argv[2], argv[3]);
    }
  else
    {
      printf("User %s, Type %s, Box: %s@%s\n",
	     argv[0], argv[1], argv[0], argv[2]);
    }
  printf("  Modified by %s on %s with %s\n",
	 argv[argc - 2], argv[argc - 3], argv[argc - 1]);
  return 0;
}

void usage(void)
{
  fprintf(stderr, "Usage: %s [-s|-S address] [-p] [-u user] [-db database]\n", whoami);
  exit(1);
}
